pkg/regserver/ampCacheregserver/ampCacheregserver.go

```diff
@@ -0,0 +1,264 @@
+	if payload == nil || clientAddrBytes == nil {
+		return
+	}
+	err := s.processor.RegisterUnidirectional(payload, pb.RegistrationSource_AMPCache, clientAddrBytes)
+
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	reqLogger.Debugf("registration successful")
+
+	// We could send an HTTP response earlier to avoid waiting
+	// while the zmq socket is locked, but this ensures that
+	// a 204 truly indicates registration success.
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (s *AMPCacheRegServer) registerBidirectional(w http.ResponseWriter, r *http.Request) {
+	s.metrics.Add("bdampcache_requests_total", 1)
+
+	logFields := log.Fields{"http_method": r.Method, "content_length": r.ContentLength, "registration_type": "bidirectional"}
+	path := strings.TrimPrefix(r.URL.Path, "/amp/register-bidirectional/")
+	reqLogger, payload, clientAddrBytes := s.registerCommon(logFields, path, w, r)
+	if payload == nil || clientAddrBytes == nil {
+		return
+	}
+
+	// Check server's client config -- add server's ClientConf if client is outdated
+	serverClientConf := s.compareClientConfGen(payload.GetRegistrationPayload().GetDecoyListGeneration())
+	if serverClientConf != nil {
+		// Replace the payload generation with correct generation from server's client config
+		payload.RegistrationPayload.DecoyListGeneration = serverClientConf.Generation
+	}
+	// Create registration response object
+	regResp, err := s.processor.RegisterBidirectional(payload, pb.RegistrationSource_BidirectionalAMP, clientAddrBytes)
+
+	if err != nil {
+		switch err {
+		case regprocessor.ErrNoC2SBody:
+			http.Error(w, "no C2S body", http.StatusBadRequest)
+		case phantoms.ErrLegacyMissingAddrs:
+			fallthrough
+		case phantoms.ErrLegacyV0SelectionBug:
+			fallthrough
+		case phantoms.ErrLegacyAddrSelectBug:
+			http.Error(w, "bad seed", http.StatusBadRequest)
+		default:
+			reqLogger.Errorf("failed to create registration response: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		return
+	}
+
+	if serverClientConf != nil {
+		// Save the client config from server to return to client
+		regResp.ClientConf = serverClientConf
+	}
+
+	w.Header().Set("Content-Type", "text/html")
+	// Attempt to hint to an AMP cache not to waste resources caching this
+	// document. "The Google AMP Cache considers any document fresh for at
+	// least 15 seconds."
+	// https://developers.google.com/amp/cache/overview#google-amp-cache-updates
+	w.Header().Set("Cache-Control", "max-age=15")
+	// Add header to w (server response)
+	w.WriteHeader(http.StatusOK)
+	// Marshal (serialize) registration response object and then write it to w
+	body, err := proto.Marshal(regResp)
+	if err != nil {
+		reqLogger.Errorf("failed to write registration into response: %v", err)
+		return
+	}
+
+	_, err = w.Write(body)
+	if err != nil {
+		reqLogger.Errorf("failed to write registration into response: %v", err)
+		return
+	}
```

Comment:
We're writing the body twice here, first without the AMP armor and then again with it. We should call `Write` only once.